libevmasm/Assembly.cpp

```diff
@@ -1465,6 +1478,17 @@ LinkerObject const& Assembly::assembleEOF() const
 		setBigEndianUint16(ret.bytecode, codeSectionSizePositions[codeSectionIndex], ret.bytecode.size() - sectionStart);
 	}
 
+	for (auto const& [refPos, tagId]: tagRef)
+	{
+		solAssert(tagId < m_tagPositionsInBytecode.size(), "Reference to non-existing tag.");
+		size_t tagPos = m_tagPositionsInBytecode[tagId];
+		solAssert(tagPos != std::numeric_limits<size_t>::max(), "Reference to tag without position.");
+
+		ptrdiff_t const relativeJumpOffset = static_cast<ptrdiff_t>(tagPos) - (static_cast<ptrdiff_t>(refPos) + 2);
+		solRequire(relativeJumpOffset < 0x7FFF && relativeJumpOffset >= -0x8000, AssemblyException, "Relative jump too far");
```

Comment:
Is `0x7FFF` really disallowed? It's still in the range of a signed 16-bit number and I don't see anything about it in the spec.

```suggestion
		solRequire(-0x8000 <= relativeJumpOffset && relativeJumpOffset <= 0x7FFF, AssemblyException, "Relative jump too far");
```

Reply:
Also, we should think what we're going to do when this happens in practice. Do we just refuse to generate code for functions that are too long?

Reply:
BTW, please also assert that the offset is not -1 or -2 since that would end up in the middle of the jump instruction itself.

Actually, it would be nice to have a more general sanity check that we're not landing in the middle of another multi-byte instruction as well, but I guess that one would be a bit complex so maybe not worth the effort. Having the offset generated from a tag should make this kind of mistake very unlikely anyway.

Reply:
We could potentially take evmone (https://github.com/ethereum/evmone/blob/master/lib/evmone/eof.cpp#L887) verification code and adjust it to our case by adding proper reporting. It can be done after bytecode generation.

Reply:
Sounds good but it would be enough as a sanity check.

Still does not solve the issue of functions that are too long. I guess that will just be a code generation error? If so, we'll need to find a way to point at the function that caused it. Errors without source location are a pain for users (we have a lot of complaints about that with StackTooDeep).